### ISSUE TITLE
Enable writing PKI certificate and private key data to a user specified Vault path(s)

### ISSUE THREAD

**Issue by allisonis** [hashicorp/vault#8486]
**Is your feature request related to a problem? Please describe.**
In order to leverage CA certificates outside of Vault, in systems that require the certificate and key to operators. Operators must go through multiple steps to safely store the private key, initial CSR, and all certificate data.
 
When generating a intermediate with Vault using the [exported type](https://www.vaultproject.io/api-docs/secret/pki/#inlinecode-type-16), the CSR and private key are returned to the caller. This information must be stored  on the machine where the request originates from. Once the CSR is generated the caller must save the CSR to a file and send it to Vault in order to have the "root" [sign the intermediate](https://www.vaultproject.io/api-docs/secret/pki/#sign-intermediate) CSR. This creates a complicated dance of managing CA certificate key material and certificate data, and attempting to safely store it. 

**Describe the solution you'd like**
When [generating an intermediate](https://www.vaultproject.io/api-docs/secret/pki/#generate-intermediate) a user can paas a flag that would re direct the output from the request, csr and private_key  or certificate data to  specified Vault path(s).

Many users or automated tooling will pull the sensitive key material from a secret path at run time. This feature would enable operators to not manually handle key material and certificates, but directly interact with Vault and improve the workflow overall. As well as reduce the risk associated with managing PKI CA Certificates, or private keys manually. 

This workflow would also work for any time a certificate is generated by the PKI backend. 

**Explain any additional use-cases**
Any PKI method that generates certificates or key material.
* Generate Certificate: [ pki/issue/:name](https://www.vaultproject.io/api-docs/secret/pki/#generate-certificate)
* Generating Intermediates: [pki/root/sign](https://www.vaultproject.io/api-docs/secret/pki/#generate-certificate)
* Sign Intermediate

**Comment by drawks** [comment]
This would be a cool feature. I will point out that you can use response wrapping to effectively do a very similar work flow to what you've detailed.

**Comment by allisonis** [comment]
Yes I have mostly created tooling to handle this limitation but I think it could improve the security and usability of the PKI backend significantly!

**Comment by allisonis** [comment]
bump

**Comment by captain-kark** [comment]
Regarding my ☝️ issue comment reference, in my case, a duplicate `transit/` endpoint containing the certificate authority's private key would be ideal. Being able to rely on continued enforcement of vault's "no export of private key matter" policy present in the pki and transit features of vault is a much better fit for me, as opposed to  using a secrets mount. I would like to set a pki mount's root certificate as "internal", as well as any intermediaries, but still be allowed the option of encrypting and decrypting arbitrary secret payloads with a single set of certificate pairs.

Should I create another issue that directly asks for the pki plugin to include support for encrypting and decrypting using the root certificate?

**Comment by HridoyRoy** [comment]
Thanks @captain-kark and @allisonis for filing this issue. I'll take a look at the newly filed issue linked above, as well as this one, and see if we can roll a joint solution out.

**Comment by anoncam** [comment]
I see this issue is open, is there a way currently in Vault managed with terraform to write that key data to a kv store path?  

The use case is PKI with [Notary](https://github.com/theupdateframework/notary)

**Comment by sgmiller** [comment]
I'm curious in everyone's use cases why the consumer who ultimately wants that private key is not the one calling the PKI secrets engine to generate it?  In the setups I've seen the receiving end (who the private key owner is talking to) is configured to accept any key whose public certificate is signed by a particular CA (Vault).  At which point there is no need for key distribution which seems like what this is trying to enable?

**Comment by bdhayco** [comment]
Just found this today and this capability is just what I require.
My organization has x.509 certs for user based entities and for our machine to machine back end comms we currently use Aethylred/keymaster in puppet to generate, store, manage and distribute keypairs to our servers. We are currently rewriting the puppet code from 3.x to 6.x and will loose this capability due to a rewrite of the Puppet CA.

I am working on a solution where Puppet will call Vault and request a Keypair be generated. But we will need to store the keypair as a KV value, should the need arise to redistribute the pair to a server.
We don't use a HSM to store keys.
My code would have had to rely on generating the key pair, and then writing the keypair back to Vault as a KV value and retrieve them as required.
Our PKI requirements cover pretty much everything, encryption of data, code signing, digital signatures, validation.
